src/coreclr/jit/rangecheck.cpp

```diff
@@ -284,8 +287,18 @@ void RangeCheck::OptimizeRangeCheck(BasicBlock* block, Statement* stmt, GenTree*
 
     if (DoesOverflow(block, treeIndex))
```

Comment:
So as I understand it the purpose of the "does overflow" checks is that RangeCheck does not model overflow (since it cannot...).

Except in the special case of all constant ranges it seems:
https://github.com/dotnet/runtime/blob/a3b186b9904844b8001d4bb5d7474f2591ee96d9/src/coreclr/jit/rangecheck.h#L238-L251

So _I think_ that the assertion "if the ranges are all constant, there could have been no overflow" is correct.

Now, what is perhaps far more curious is the "does it overflow" logic itself, which says that if the end result of the SSA use-def chain walk is an `IND`, it does not overflow. This makes sense insofar as RangeCheck does not work with indirections, thus could not have drawn wrong conclusions earlier, when computing the ranges (another way to put it is that an indirection in the dependency chain will always be terminating, not result in a dependent range).

What makes me wonder then is why don't we consider chains ending in, say, parameters (as is the case here) as non-overflowing.

Reply:
Thanks for looking at it, I agree it's a bit confusing